index.ts

```diff
@@ -0,0 +1,74 @@
+import { EventEmitter } from 'events';
+
+type Handler = (...args: any[]) => void;
+type EventMap = {
+  [k: string]: Handler | Handler[] | undefined;
+};
+
+function safeApply<T, A extends any[]> (handler: (this: T, ...args: A) => void, context: T, args: A): void {
+  try {
+    Reflect.apply(handler, context, args);
+  } catch (err) {
+    // Throw error after timeout so as not to interrupt the stack
+    setTimeout(() => {
+      throw err;
+    });
+  }
+}
+
+function arrayClone<T> (arr: T[]): T[] {
+  const n = arr.length;
+  const copy = new Array(n);
+  for (let i = 0; i < n; i++) {
+    copy[i] = arr[i];
+  }
+  return copy;
+}
+
+export default class SafeEventEmitter extends EventEmitter {
+  emit (type: string, ...args: any[]): boolean {
+    let doError = type === 'error';
+
+    const events: EventMap = (this as any)._events;
+    if (events !== undefined) {
+      doError = doError && events.error === undefined;
+    } else if (!doError) {
+      return false;
+    }
+
+    // If there is no 'error' event listener then throw.
+    if (doError) {
+      let er;
+      if (args.length > 0) {
+        er = args[0];
+      }
+      if (er instanceof Error) {
+        // Note: The comments on the `throw` lines are intentional, they show
+        // up in Node's output if this results in an unhandled exception.
+        throw er; // Unhandled 'error' event
+      }
+      // At least give some kind of context to the user
+      const err = new Error('Unhandled error.' + (er ? ' (' + er.message + ')' : ''));
+      (err as any).context = er;
+      throw err; // Unhandled 'error' event
+    }
+
+    const handler = events[type];
+
+    if (handler === undefined) {
+      return false;
+    }
+
+    if (typeof handler === 'function') {
+      safeApply(handler, this, args);
+    } else {
+      const len = handler.length;
+      const listeners = arrayClone(handler);
```

Comment:
Actually.... this is the only place where `arrayClone` is used, and it's entirely pointless? It makes a shallow clone, then discards it. The `listeners` array is never touched after this point - just `listeners[i]`, which is the same as `handler[i]` because shallow. So `arrayClone` can be removed altogether.

Reply:
Huh - apparently this is in the `Gozala/events` module as well.

Reply:
It's also in node core. It was added here: https://github.com/nodejs/node/pull/601
I must be missing something 🤔 I don't understand why that clone is happening at all, and from skimming through the discussion they don't mention it at all. Maybe some kinda obscure micro-optimization 🤷‍♂ .

Reply:
Yeah even @rekmarks' suggestions above, I assume they were all micro-optimizations in (what could be) hot code paths. I left in the `len` variable here for that reason. A part of me wants to clean everything up and just say "let the runtime deal with it," but... yeah. 🤷‍♂ 

That _was_ 5 years ago, and runtimes have come a long way since. 🤔

Reply:
I was poking around on jsperf for that but couldn't find anything conclusive. The good news is, it's definitely fine, and there's no reason to block on it. I don't see any reason not to do `[ ...handler ]`, but 🤷‍♂

Reply:
It turns out the clone is necessary 🤦‍♂ . Kinda obvious in retrospect - the clone happens here because the event handlers might subscribe new handlers, or unsubscribe them, thus mutating the handlers array.